### ISSUE TITLE
udev does not load module if modalias contains a single quote '\''

### ISSUE THREAD

**Issue by jwrdegoede** [systemd/systemd#24715]
### systemd version the issue has been seen with

251.4

### Used distribution

Fedora 37

### Linux kernel version used

5.19.? and 6.0.0-rc?

### CPU architectures issue was seen on

x86_64

### Component

systemd-udevd

### Expected behaviour you didn't see

msi-laptop module does not get autoloaded when it should.

### Steps to reproduce the problem

On a MSI S270 laptop the dmi modalias is:

 ```
# cat /sys/class/dmi/id/modalias
dmi:bvnAmericanMegatrendsInc.:bvrA1013AMSV4.30:bd01/09/2006:svnMICRO-STARINT'LCO.,LTD:pnMS-1013:pvr0131:cvnMICRO-STARINT'LCO.,LTD:ct10:cvrN/A:
```

And one of the module aliases for the msi-laptop kernel module is:

```
dmi*:svn*MICRO-STARINT'LCO.,LTD*:pn*MS-1013*:pvr*0131*:cvn*MICRO-STARINT'LCO.,LTD*:
```

Yet the module does not autoload, while:

```
modprobe `cat /sys/class/dmi/id/modalias`
```

does cause it to autoload.

On a hunch I dropped the `'LCO.,LTD` part of the DMI strings in the kernel module, so that the module modalias now is:

```
dmi*:svn*MICRO-STARINT*:pn*MS-1013*:pvr*0131*:cvn*MICRO-STARINT*:
```

And that works around this issue.

I have the feeling that udevd is doing something wrong with escaping the `'` char, perhaps passing something like:
```
dmi:bvnAmericanMegatrendsInc.:bvrA1013AMSV4.30:bd01/09/2006:svnMICRO-STARINT\'LCO.,LTD:pnMS-1013:pvr0131:cvnMICRO-STARINT\'LCO.,LTD:ct10:cvrN/A:
```

to the modprobe command. Notice the backslashes there which should not be there. These extra backslashes are causing the msi-laptop alias to not match the modalias (as passed to modprobe). By changing the msi-laptop alias to instead of `'LCO.,LTD` accepting a wildcard (`*`) in that part of the alias the extra escaping does not matter and the module does auto-load correctly.

Note this addition of extra blackslashes to escape the `'` is only a theory I have not confirmed this is actually happening in any way.

**Comment by yuwata** [comment]
Could you show `MODALIAS` property shown by `udevadm info /sys/class/dmi/id`?

**Comment by jwrdegoede** [comment]
```
[root@fedora ~]# udevadm info /sys/class/dmi/id
P: /devices/virtual/dmi/id
M: id
U: dmi
E: DEVPATH=/devices/virtual/dmi/id
E: SUBSYSTEM=dmi
E: MODALIAS=dmi:bvnAmericanMegatrendsInc.:bvrA1013AMSV4.30:bd01/09/2006:svnMICRO-STARINT'LCO.,LTD:pnMS-1013:pvr0131:cvnMICRO-STARINT'LCO.,LTD:ct10:cvrN/A:

[root@fedora ~]# cat /sys/class/dmi/id/uevent
MODALIAS=dmi:bvnAmericanMegatrendsInc.:bvrA1013AMSV4.30:bd01/09/2006:svnMICRO-STARINT'LCO.,LTD:pnMS-1013:pvr0131:cvnMICRO-STARINT'LCO.,LTD:ct10:cvrN/A:
```

Here you go (note this was already in my original report in the form of `cat /sys/class/dmi/id/modalias` output).

**Comment by yuwata** [comment]
Note, udevd wrongly pass modalias `dmi:bvnAmericanMegatrendsInc.:bvrA1013AMSV4.30:bd01/09/2006:svnMICRO-STARINTLCO.,LTD:pnMS-1013:pvr0131:cvnMICRO-STARINTLCO.,LTD:ct10:cvrN/A:` to kmod. That is, all quotation is dropped.

**Comment by jwrdegoede** [comment]
> Note, udevd wrongly(?) pass modalias `dmi:bvnAmericanMegatrendsInc.:bvrA1013AMSV4.30:bd01/09/2006:svnMICRO-STARINTLCO.,LTD:pnMS-1013:pvr0131:cvnMICRO-STARINTLCO.,LTD:ct10:cvrN/A:` to kmod. That is, all quotation is dropped.

Interesting thank you for figuring that out.

**Comment by yuwata** [comment]
More backgrounds.

The relevant udev rule is https://github.com/systemd/systemd/blob/cf610e1d873d6bf1a6017597ef50a6cf68f18313/rules.d/80-drivers.rules#L5
The quote in command is dropped before executed, https://github.com/systemd/systemd/blob/cf610e1d873d6bf1a6017597ef50a6cf68f18313/src/udev/udev-builtin.c#L112

As you can see, e.g. `echo 'foo'a'bbb` outputs `fooabbb`, so such dropping quote within the argument sounds 'natural' and existing udev rules may already expect that behavior.

Though, that's builtin command, not generic `RUN=` rule. If so many hardware already have quote in modalias, then maybe we should change the way to split the command to drop the quote only at the beginning and end of each argument.

**Comment by yuwata** [comment]
From https://github.com/linuxhw/DMI, I found the followings which contain quote in DMI (though, these are from dmidecode, not sure about modalias): 
- Smart'MOUV series
https://github.com/linuxhw/DMI/blob/8d962e653f557f435d7b8aac6111de6f17535336/Desktop/essentielb/SmartMOUV/SmartMOUV%20series/067A94E9A34E#L40
- many MSI products, e.g.
https://github.com/linuxhw/DMI/blob/8d962e653f557f435d7b8aac6111de6f17535336/Notebook/LG%20Electronics/E500/E500-U.AP21R/C845B4FEA5B4#L48
- KREZ TM1004 Tap'n'Go
https://github.com/linuxhw/DMI/blob/8d962e653f557f435d7b8aac6111de6f17535336/Notebook/Insyde/KREZ/KREZ%20TM1004%20Tap'n'Go/F8EED61AACCD#L38
- Let's note
https://github.com/linuxhw/DMI/blob/8d962e653f557f435d7b8aac6111de6f17535336/Notebook/Panasonic/CFSV9/CFSV9-2/E098ED5E2EC3#L38
- CompuLab AMD "CM-iGLX" Geode LX/CS5536
https://github.com/linuxhw/DMI/blob/8d962e653f557f435d7b8aac6111de6f17535336/Desktop/CompuLab/AMD/AMD%20%22CM-iGLX%22%20Geode%20LX-CS5536/7EF0E57B46AC#L38

**Comment by yuwata** [comment]
@jwrdegoede 'Fix' is waiting in #24998.
